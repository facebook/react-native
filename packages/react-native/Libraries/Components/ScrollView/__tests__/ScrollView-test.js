/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow-strict
 * @format
 * @oncall react_native
 */

'use strict';

const {create, unmount, update} = require('../../../../jest/renderer');
const Text = require('../../../Text/Text');
const ReactNativeTestTools = require('../../../Utilities/ReactNativeTestTools');
const View = require('../../View/View');
const ScrollView = require('../ScrollView');
const React = require('react');

describe('ScrollView', () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it('renders its children', async () => {
    await ReactNativeTestTools.expectRendersMatchingSnapshot(
      'ScrollView',
      () => (
        <ScrollView>
          <View>
            <Text>Hello World!</Text>
          </View>
        </ScrollView>
      ),
      () => {
        jest.dontMock('../ScrollView');
      },
    );
  });

  it('mocks native methods and instance methods', async () => {
    jest.mock('../ScrollView');

    const ref = React.createRef();
    await create(<ScrollView ref={ref} />);

    expect(ref.current?.measure).toBeInstanceOf(jest.fn().constructor);
    expect(ref.current?.scrollTo).toBeInstanceOf(jest.fn().constructor);
  });

  describe('ref', () => {
    it('receives an instance or null', async () => {
      jest.dontMock('../ScrollView');

      const scrollViewRef = jest.fn();
      const testRendererInstance = await create(
        <ScrollView ref={scrollViewRef} />,
      );

      expect(scrollViewRef).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );

      await unmount(testRendererInstance);

      expect(scrollViewRef).toHaveBeenLastCalledWith(null);
    });

    it('transitions between refs', async () => {
      jest.dontMock('../ScrollView');

      const scrollViewRefA = jest.fn();
      const testRendererInstance = await create(
        <ScrollView ref={scrollViewRefA} />,
      );

      expect(scrollViewRefA).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );

      const scrollViewRefB = jest.fn();
      await update(testRendererInstance, <ScrollView ref={scrollViewRefB} />);

      expect(scrollViewRefA).toHaveBeenLastCalledWith(null);
      expect(scrollViewRefB).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );
    });
  });

  describe('innerViewRef', () => {
    it('receives an instance or null', async () => {
      jest.dontMock('../ScrollView');

      const innerViewRef = jest.fn();
      const testRendererInstance = await create(
        <ScrollView innerViewRef={innerViewRef} />,
      );

      expect(innerViewRef).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );

      await unmount(testRendererInstance);

      expect(innerViewRef).toHaveBeenLastCalledWith(null);
    });

    it('transitions between refs', async () => {
      jest.dontMock('../ScrollView');

      const innerViewRefA = jest.fn();
      const testRendererInstance = await create(
        <ScrollView innerViewRef={innerViewRefA} />,
      );

      expect(innerViewRefA).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );

      const innerViewRefB = jest.fn();

      await update(
        testRendererInstance,
        <ScrollView innerViewRef={innerViewRefB} />,
      );

      expect(innerViewRefA).toHaveBeenLastCalledWith(null);
      expect(innerViewRefB).toHaveBeenLastCalledWith(
        expect.objectContaining({_nativeTag: expect.any(Number)}),
      );
    });
  });

  describe('getInnerViewRef', () => {
    it('returns an instance', async () => {
      jest.dontMock('../ScrollView');

      const scrollViewRef = React.createRef(null);
      await create(<ScrollView ref={scrollViewRef} />);
      const innerViewRef = scrollViewRef.current.getInnerViewRef();

      // This is checking if the ref acts like a host component. If we had an
      // `isHostComponent(ref)` method, that would be preferred.
      expect(innerViewRef.measure).toBeInstanceOf(jest.fn().constructor);
      expect(innerViewRef.measureLayout).toBeInstanceOf(jest.fn().constructor);
      expect(innerViewRef.measureInWindow).toBeInstanceOf(
        jest.fn().constructor,
      );
    });
  });
});
