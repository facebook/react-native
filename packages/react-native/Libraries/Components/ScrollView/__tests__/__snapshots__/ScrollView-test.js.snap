// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollView renders its children: should deep render when mocked (please verify output manually) 1`] = `
<RCTScrollView>
  <View>
    <View>
      <Text>
        Hello World!
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ScrollView renders its children: should deep render when not mocked (please verify output manually) 1`] = `
<RCTScrollView
  alwaysBounceVertical={true}
  onContentSizeChange={null}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminationRequest={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onScrollShouldSetResponder={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  pagingEnabled={false}
  sendMomentumEvents={false}
  snapToEnd={true}
  snapToStart={true}
  style={
    Object {
      "flexDirection": "column",
      "flexGrow": 1,
      "flexShrink": 1,
      "overflow": "scroll",
    }
  }
>
  <RCTScrollContentView
    collapsable={false}
    collapsableChildren={true}
    style={
      Array [
        false,
        undefined,
      ]
    }
  >
    <View>
      <Text>
        Hello World!
      </Text>
    </View>
  </RCTScrollContentView>
</RCTScrollView>
`;
